gateway/handle.go

```diff
@@ -684,73 +695,71 @@ func (gw *Handle) internalBatchHandlerFunc() http.HandlerFunc {
+			logger.NewStringField("path", r.URL.Path),
+			logger.NewIntField("status", int64(status)),
+			logger.NewStringField("body", responseBody),
+			logger.NewStringField("request", string(body)),
+		)
 		http.Error(w, responseBody, status)
 	}
 }
 
-func (gw *Handle) extractJobsFromInternalBatchPayload(
-	arctx *gwtypes.AuthRequestContext,
-	reqType string,
-	body []byte,
-) ([]*jobsdb.JobT, error) {
-	if !gjson.ValidBytes(body) {
-		return nil, fmt.Errorf("%s", response.InvalidJSON)
+func (gw *Handle) extractJobsFromInternalBatchPayload(reqType string, body []byte) ([]*jobsdb.JobT, error) {
+	type params struct {
+		MessageID       string `json:"message_id"`
+		SourceID        string `json:"source_id"`
+		SourceJobRunID  string `json:"source_job_run_id"`
+		SourceTaskRunID string `json:"source_task_run_id"`
+		UserID          string `json:"user_id"`
+		TraceParent     string `json:"traceparent"`
 	}
-	gw.requestSizeStat.Observe(float64(len(body)))
 
-	type jobObject struct {
-		userID     string
-		events     []map[string]interface{}
-		receivedAt string
+	type singularEventBatch struct {
+		Batch      []json.RawMessage `json:"batch"`
+		ReceivedAt string            `json:"receivedAt"`
+		RequestIP  string            `json:"requestIP"`
 	}
+
 	var (
-		sourcesJobRunID  = arctx.SourceJobRunID
-		sourcesTaskRunID = arctx.SourceTaskRunID
-		sourceID         = arctx.SourceID
-		workspaceID      = arctx.WorkspaceID
-		eventsBatch      = gjson.GetBytes(body, "batch").Array()
+		messages         []stream.Message
 		isUserSuppressed = gw.memoizedIsUserSuppressed()
-		out              = make([]jobObject, 0, len(eventsBatch))
+		jobs             []*jobsdb.JobT
 	)
 
-	for idx, v := range eventsBatch {
-		toSet, ok := v.Value().(map[string]interface{})
-		if !ok {
-			gw.stats.NewTaggedStat(
-				"gateway.write_key_failed_events",
-				stats.CountType,
-				gw.newSourceStatTagsWithReason(arctx, reqType, response.NotRudderEvent),
-			).Increment()
-			return nil, fmt.Errorf("%s", response.NotRudderEvent)
-		}
-		anonIDFromReq, _ := toSet["anonymousId"].(string)
-		userIDFromReq, _ := toSet["userId"].(string)
-		eventContext, ok := misc.MapLookup(toSet, "context").(map[string]interface{})
-		if ok {
-			if idx == 0 {
-				if v, _ := misc.MapLookup(eventContext, "sources", "job_run_id").(string); v != "" {
-					sourcesJobRunID = v
-				}
-				if v, _ := misc.MapLookup(eventContext, "sources", "task_run_id").(string); v != "" {
-					sourcesTaskRunID = v
-				}
-			}
-		}
+	err := jsonfast.Unmarshal(body, &messages)
+	if err != nil {
+		return nil, errors.New(response.InvalidJSON)
+	}
+	gw.requestSizeStat.Observe(float64(len(body)))
 
-		if isUserSuppressed(workspaceID, userIDFromReq, sourceID) {
+	if len(messages) == 0 {
+		return nil, errors.New(response.NotRudderEvent)
+	}
+
+	jobs = make([]*jobsdb.JobT, 0, len(messages))
+
+	for _, msg := range messages {
+		err := gw.streamMsgValidator(&msg)
+		if err != nil {
```

Comment:
Shall we log the error here, we are losing information since we are returning just `errors.New(response.InvalidStreamMessage)` 
```suggestion
		if err != nil {
```